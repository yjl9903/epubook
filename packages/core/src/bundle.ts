import * as fflate from 'fflate';
import { XMLBuilder } from 'fast-xml-parser';

import type { Epubook } from './epub';

const MIMETYPE = fflate.strToU8('application/epub+zip');

/**
 * Bundle epub to zip archive
 *
 * @param epub
 * @returns
 */
export async function bundle(epub: Epubook) {
  return new Promise((res, rej) => {
    fflate.zip(
      {
        mimetype: MIMETYPE,
        'META-INF': {
          'container.xml': fflate.strToU8(makeContainer(epub))
        },
        OEBPS: {
          'content.opf': fflate.strToU8('')
        }
      },
      {
        level: 0,
        mtime: new Date()
      },
      (err, data) => {
        if (err) {
          rej(err);
        } else {
          res(data);
        }
      }
    );
  });
}

/**
 * Generate META-INF/container.xml
 *
 * See: https://www.w3.org/TR/epub-33/#sec-container-metainf-container.xml
 *
 * Example: https://www.w3.org/TR/epub-33/#sec-container-container.xml-example
 *
 * @param epub
 * @returns xml string
 */
export function makeContainer(epub: Epubook): string {
  const builder = new XMLBuilder({
    format: true,
    ignoreAttributes: false,
    suppressUnpairedNode: false,
    unpairedTags: ['rootfile']
  });

  return builder.build({
    '?xml': { '#text': '', '@_version': '1.0', '@_encoding': 'UTF-8' },
    container: {
      '@_version': '1.0',
      '@_xmlns': 'urn:oasis:names:tc:opendocument:xmlns:container',
      rootfiles: [
        {
          rootfile: [
            {
              '@_full-path': 'OEBPS/content.opf',
              '@_media-type': 'application/oebps-package+xml',
              '#text': ''
            }
          ]
        }
      ]
    }
  });
}
