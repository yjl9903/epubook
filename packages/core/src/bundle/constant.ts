export const MIMETYPE = 'application/epub+zip';

export const ImageGif = 'image/gif';
export const ImageJpeg = 'image/jpeg';
export const ImagePng = 'image/png';
export const ImageSvg = 'image/svg+xml';
export const ImageWebp = 'image/webp';

export type ImageMediaType = typeof ImageGif | typeof ImageJpeg | typeof ImagePng;

export const TextCSS = 'text/css';

export const XHTML = 'application/xhtml+xml';

export type MediaType = ImageMediaType | typeof TextCSS | typeof XHTML;
