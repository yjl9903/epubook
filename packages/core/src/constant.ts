import * as path from 'node:path';

export const MIMETYPE = 'application/epub+zip';

export const ImageGif = 'image/gif';
export const ImageJpeg = 'image/jpeg';
export const ImagePng = 'image/png';
export const ImageSvg = 'image/svg+xml';
export const ImageWebp = 'image/webp';

export type ImageMediaType =
  | typeof ImageGif
  | typeof ImageJpeg
  | typeof ImagePng
  | typeof ImageSvg
  | typeof ImageWebp;

export type ImageExtension = 'gif' | 'jpg' | 'jpeg' | 'png' | 'svg' | 'webp';

export const TextCSS = 'text/css';

export const XHTML = 'application/xhtml+xml';

export type MediaType = ImageMediaType | typeof TextCSS | typeof XHTML;

export function getImageMediaType(file: string): ImageMediaType | undefined {
  const ext = path.extname(file);
  switch (ext) {
    case '.gif':
      return ImageGif;
    case '.jpg':
    case '.jpeg':
      return ImageJpeg;
    case '.png':
      return ImagePng;
    case '.svg':
      return ImageSvg;
    case '.webp':
      return ImageWebp;
    default:
      return undefined;
  }
}
