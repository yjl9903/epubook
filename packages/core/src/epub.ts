import { randomUUID } from 'node:crypto';
import { type PathLike, promises as fs } from 'node:fs';

import { createDefu } from 'defu';

export class Epubook {
  /**
   * See: https://www.w3.org/TR/epub-33/#sec-package-doc
   *
   * Now, it only supports single opf (OEBPS/content.opf)
   *
   * @returns list of package documents
   */
  private opfs: PackageDocument[] = [new PackageDocument('OEBPS/content.opf')];

  constructor() {}

  public packageDocuments(): PackageDocument[] {
    return this.opfs;
  }

  public mainPackageDocument() {
    return this.opfs[0];
  }

  async bundle() {
    const { bundle } = await import('./bundle');
    return await bundle(this);
  }

  async writeFile(file: PathLike) {
    const buffer = await this.bundle();
    await fs.writeFile(file, buffer);
  }
}

const defu = createDefu((obj: any, key, value: any) => {
  if (obj[key] instanceof Date && value instanceof Date) {
    obj[key] = value;
    return true;
  }
});

export class PackageDocument {
  private readonly file: string;

  private readonly SpecVersion: '3.0' = '3.0';

  private _uniqueIdentifier = 'uuid';

  private _identifier = randomUUID();

  private _metadata = {
    title: '',
    language: 'zh-CN',
    contributor: [] as string[],
    coverage: '',
    creator: 'unknown',
    date: new Date(),
    description: '',
    format: '',
    publisher: '',
    relation: '',
    rights: '',
    source: '',
    subject: '',
    type: '',
    lastModified: new Date()
  };

  constructor(file: string) {
    this.file = file;
  }

  public filename() {
    return this.file;
  }

  public version() {
    return this.SpecVersion;
  }

  // --- metadata ---
  public update(info: Partial<typeof this._metadata>) {
    // TODO: valiate input data
    this._metadata = defu(info, this._metadata);
    return this;
  }

  public title() {
    return this._metadata.title;
  }

  public language() {
    return this._metadata.language;
  }

  public creator() {
    return this._metadata.creator;
  }

  public metadata() {
    return this._metadata;
  }

  // --- identifier ---
  public uniqueIdentifier() {
    return this._uniqueIdentifier;
  }

  public identifier() {
    return this._identifier;
  }

  public setIdentifier(identifier: string, uniqueIdentifier: string) {
    this._identifier = identifier;
    this._uniqueIdentifier = uniqueIdentifier;
  }
}
