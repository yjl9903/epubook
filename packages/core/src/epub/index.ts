export * from './epub';

export * from './opf';
