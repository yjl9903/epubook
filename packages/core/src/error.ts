export class BundleError extends Error {
  constructor(msg: string) {
    super(msg);
  }
}
