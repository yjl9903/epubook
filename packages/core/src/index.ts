export * from './epub';

export * from './theme';

export * from './constant';

export * from '@epubook/xml';
