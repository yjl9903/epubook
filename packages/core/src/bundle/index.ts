import * as path from 'pathe';
import * as fflate from 'fflate';
import { XMLBuilder } from 'fast-xml-parser';

import type { Epubook, ManifestItem, ManifestItemRef, PackageDocument } from '../epub';

import { MIMETYPE } from '../constant';
import { BundleError } from '../error';
import { toISO8601String } from '../utils';

/**
 * Bundle epub to zip archive
 *
 * @param epub
 * @returns
 */
export async function bundle(epub: Epubook): Promise<Uint8Array> {
  return new Promise(async (res, rej) => {
    const opfs = epub
      .packageDocuments()
      .map((opf) => [opf.filename(), fflate.strToU8(makePackageDocument(opf))] as const);

    const items: Record<string, Uint8Array> = {};
    for (const opf of epub.packageDocuments()) {
      const base = path.dirname(opf.filename());
      for (const item of opf.items()) {
        const name = path.join(base, item.filename());
        if (name in items) {
          continue;
        }
        // TODO: parallel here
        items[name] = await item.bundle();
      }
    }

    const abstractContainer: fflate.AsyncZippable = {
      mimetype: fflate.strToU8(MIMETYPE),
      'META-INF': {
        'container.xml': fflate.strToU8(makeContainer(epub))
      },
      ...Object.fromEntries(opfs),
      ...items
    };

    fflate.zip(
      abstractContainer,
      {
        level: 0,
        mtime: new Date()
      },
      (err, data) => {
        if (err) {
          rej(err);
        } else {
          res(data);
        }
      }
    );
  });
}

/**
 * Generate META-INF/container.xml
 *
 * See: https://www.w3.org/TR/epub-33/#sec-container-metainf-container.xml
 *
 * Example: https://www.w3.org/TR/epub-33/#sec-container-container.xml-example
 *
 * @param epub
 * @returns xml string
 */
export function makeContainer(epub: Epubook): string {
  const builder = new XMLBuilder({
    format: true,
    ignoreAttributes: false,
    suppressUnpairedNode: false,
    unpairedTags: ['rootfile']
  });

  const rootfile = epub.packageDocuments().map((p) => ({
    '@_full-path': p.filename(),
    '@_media-type': 'application/oebps-package+xml',
    '#text': ''
  }));

  return builder.build({
    '?xml': { '#text': '', '@_version': '1.0', '@_encoding': 'UTF-8' },
    container: {
      '@_version': '1.0',
      '@_xmlns': 'urn:oasis:names:tc:opendocument:xmlns:container',
      rootfiles: [
        {
          rootfile
        }
      ]
    }
  });
}

/**
 * Generate package document
 *
 * @param opf
 * @returns
 */
export function makePackageDocument(opf: PackageDocument): string {
  if (opf.version() !== '3.0') {
    throw new BundleError(`Unsupport EPUB spec ${opf.version()}`);
  }

  const builder = new XMLBuilder({
    format: true,
    ignoreAttributes: false,
    suppressUnpairedNode: false,
    unpairedTags: ['item', 'itemref']
  });

  const optionalMetadata: Record<string, any> = {};
  const optionalList: Array<keyof ReturnType<typeof opf.metadata>> = [
    'contributor',
    'coverage',
    'format',
    'publisher',
    'relation',
    'rights',
    'source',
    'subject',
    'type'
  ];
  for (const key of optionalList) {
    const m = opf.metadata();
    if (!!m[key]) {
      optionalMetadata['dc:' + key] = m[key];
    }
  }
  const metadata = {
    '@_xmlns:dc': 'http://purl.org/dc/elements/1.1/',
    'dc:identifier': {
      '@_id': opf.uniqueIdentifier(),
      '#text': opf.identifier()
    },
    'dc:title': opf.title(),
    'dc:language': opf.language(),
    'dc:creator': {
      '@_id': 'creator',
      '#text': opf.creator()
    },
    'dc:date': toISO8601String(opf.metadata().date),
    'dc:description': opf.metadata().description,
    ...optionalMetadata,
    meta: [
      {
        '@_property': 'dcterms:modified',
        '#text': toISO8601String(opf.metadata().lastModified)
      },
      {
        '@_refines': '#creator',
        '@_property': 'file-as',
        '#text': opf.creator()
      }
    ]
  };

  function makeManifestItem(item: ManifestItem) {
    return {
      '@_fallback': item.fallback(),
      '@_href': item.href(),
      '@_id': item.id(),
      '@_media-overlay': item.mediaOverlay(),
      '@_media-type': item.mediaType(),
      '@_properties': item.properties()
    };
  }

  function makeManifestItemRef(item: ManifestItemRef) {
    return {
      '@_idref': item.idref()
    };
  }

  return builder.build({
    '?xml': { '#text': '', '@_version': '1.0', '@_encoding': 'UTF-8' },
    package: {
      '@_xmlns': 'http://www.idpf.org/2007/opf',
      '@_xmlns:epub': 'http://www.idpf.org/2007/ops',
      '@_unique-identifier': opf.uniqueIdentifier(),
      '@_version': opf.version(),
      metadata,
      manifest: {
        item: opf.manifest().map((i) => makeManifestItem(i))
      },
      spine: {
        itemref: opf.spine().map((ir) => makeManifestItemRef(ir))
      }
    }
  });
}
