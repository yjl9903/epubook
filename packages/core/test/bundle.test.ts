import { describe, it, expect } from 'vitest';

import { Epub, Toc, HTML, XHTMLBuilder } from '../src';
import { makeContainer, makePackageDocument } from '../src/bundle';

describe('Bundle Epub', () => {
  it('generate container.xml', () => {
    const res = makeContainer(new Epub());
    expect(res).toMatchInlineSnapshot(`
      "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
      <container version=\\"1.0\\" xmlns=\\"urn:oasis:names:tc:opendocument:xmlns:container\\">
        <rootfiles>
          <rootfile full-path=\\"OEBPS/content.opf\\" media-type=\\"application/oebps-package+xml\\"/>
        </rootfiles>
      </container>
      "
    `);
  });

  it('generate opf', () => {
    const epub = new Epub();

    const opf = epub.main();
    opf.setIdentifier('test-book-id', 'BookId');
    opf.update({
      title: 'Test Book',
      language: 'zh-CN',
      date: new Date('2023-02-01T11:00:00.000Z'),
      lastModified: new Date('2023-02-26T11:00:00.000Z'),
      creator: { name: 'XLor', fileAs: 'XLor' },
      description: 'for test usage',
      source: 'imagine'
    });

    const cover = new HTML('cover.xhtml', '');
    epub.item(cover).spine(cover);

    const res = makePackageDocument(opf);
    expect(res).toMatchInlineSnapshot(`
      "<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
      <package xmlns=\\"http://www.idpf.org/2007/opf\\" xmlns:epub=\\"http://www.idpf.org/2007/ops\\" unique-identifier=\\"BookId\\" version=\\"3.0\\">
        <metadata xmlns:dc=\\"http://purl.org/dc/elements/1.1/\\">
          <dc:identifier id=\\"BookId\\">test-book-id</dc:identifier>
          <dc:title>Test Book</dc:title>
          <dc:language>zh-CN</dc:language>
          <dc:creator id=\\"creator\\">XLor</dc:creator>
          <dc:date>2023-02-01T11:00:00Z</dc:date>
          <dc:description>for test usage</dc:description>
          <dc:source>imagine</dc:source>
          <meta property=\\"dcterms:modified\\">2023-02-26T11:00:00Z</meta>
          <meta refines=\\"#creator\\" property=\\"file-as\\">XLor</meta>
        </metadata>
        <manifest>
          <item href=\\"cover.xhtml\\" id=\\"cover\\" media-type=\\"application/xhtml+xml\\"/>
        </manifest>
        <spine>
          <itemref idref=\\"cover\\"/>
        </spine>
      </package>
      "
    `);
  });

  it('write epub', async () => {
    const epub = new Epub({
      title: 'Test Book',
      date: new Date('2023-02-01T11:00:00.000Z'),
      lastModified: new Date('2023-02-26T11:00:00.000Z'),
      creator: { name: 'XLor', fileAs: 'XLor' },
      description: 'for test usage',
      source: 'imagine'
    });

    epub.main().setIdentifier('test-book-id', 'BookId');

    const content = `<html xmlns="http://www.w3.org/1999/xhtml" xmlns:epub="http://www.idpf.org/2007/ops" xml:lang="en">
  <head>
    <title>Data URL does not open in top-level context</title>
    <style>
      code {
        color: #c63501;
      }
    </style>
  </head>
  <body>
    <h1>Data URL does not open in top-level context</h1>
    <p>The following jpeg is contained within a <code>data:</code> URL, which is used as the <code>src</code> attribute for an <code>img</code> element.</p>
    <p>The test passes if you are able to see the image below inside this ebook.</p>
  </body>
</html>`;
    const cover = new HTML('cover.xhtml', content);
    epub
      .item(cover)
      .spine(cover)
      .toc([{ title: 'cover', page: cover }]);

    await epub.writeFile('../../.output/test-core.epub');
  });

  it('generate toc', async () => {
    const epub = new Epub({
      title: 'Test Book',
      date: new Date('2023-02-01T11:00:00.000Z'),
      lastModified: new Date('2023-02-26T11:00:00.000Z'),
      creator: {
        name: 'XLor',
        fileAs: 'XLor'
      },
      description: 'for test usage',
      source: 'imagine'
    });
    epub.main().setIdentifier('12345', 'book-id');

    const item1 = new HTML('page1.xhtml', '1');
    const item2 = new HTML('page2.xhtml', '2');
    const item3 = new HTML('page3.xhtml', '3');
    const item4 = new HTML('page4.xhtml', '4');
    const item5 = new HTML('page5.xhtml', '5');

    await epub
      .item(item1, item2, item3, item4, item5)
      .spine(item1, item2, item3, item4, item5)
      .toc(
        [
          { title: '1', page: item1 },
          { title: '2', page: item2 },
          {
            title: 'Sub',
            list: [
              { title: '3', page: item3 },
              { title: '4', page: item4 }
            ]
          },
          { title: '5', page: item5 }
        ],
        {
          heading: 2,
          title: 'Toc'
        }
      )
      .bundle();

    expect(epub.main()).toMatchSnapshot();
  });
});

describe('XHTML Builder', () => {
  it('should build empty', () => {
    const builder = new XHTMLBuilder('a.xhtml');
    const res = builder.build();
    expect(res).toMatchInlineSnapshot(`
      XHTML {
        "_content": "<html xmlns=\\"http://www.w3.org/1999/xhtml\\" xmlns:epub=\\"http://www.idpf.org/2007/ops\\" lang=\\"en\\" xml:lang=\\"en\\">
        <head>
          <title>a.xhtml</title>
        </head>
        <body></body>
      </html>
      ",
        "_meta": {
          "language": "en",
          "title": "a.xhtml",
        },
        "_properties": undefined,
        "file": "a.xhtml",
        "mediaType": "application/xhtml+xml",
      }
    `);
  });

  it('should build styles', () => {
    const builder = new XHTMLBuilder('a.xhtml');
    const res = builder.title('with style').style('123').style('456').build();
    expect(res).toMatchInlineSnapshot(`
      XHTML {
        "_content": "<html xmlns=\\"http://www.w3.org/1999/xhtml\\" xmlns:epub=\\"http://www.idpf.org/2007/ops\\" lang=\\"en\\" xml:lang=\\"en\\">
        <head>
          <title>with style</title>
          <link href=\\"123\\" rel=\\"stylesheet\\" type=\\"text/css\\"/>
          <link href=\\"456\\" rel=\\"stylesheet\\" type=\\"text/css\\"/>
        </head>
        <body></body>
      </html>
      ",
        "_meta": {
          "language": "en",
          "title": "with style",
        },
        "_properties": undefined,
        "file": "a.xhtml",
        "mediaType": "application/xhtml+xml",
      }
    `);
  });

  it('should build nav toc', () => {
    const toc = new Toc('nav.xhtml');
    const page1 = new HTML('a.xhtml', '');
    const page2 = new HTML('b.xhtml', '');
    const page4 = new HTML('d.xhtml', '');
    const res = toc
      .generate([
        { title: 'page 1', page: page1 },
        { title: 'page 2', page: page2 },
        { title: 'page 3', list: [{ title: 'page 4', page: page4 }] }
      ])
      .build();

    expect(res).toMatchInlineSnapshot(`
      XHTML {
        "_content": "<html xmlns=\\"http://www.w3.org/1999/xhtml\\" xmlns:epub=\\"http://www.idpf.org/2007/ops\\" lang=\\"en\\" xml:lang=\\"en\\">
        <head>
          <title>Nav</title>
        </head>
        <body>
          <nav epub:type=\\"toc\\">
            <h1>Nav</h1>
            <ol>
              <li>
                <a href=\\"a.xhtml\\">page 1</a>
              </li>
              <li>
                <a href=\\"b.xhtml\\">page 2</a>
              </li>
              <li>
                <span>page 3</span>
                <ol>
                  <li>
                    <a href=\\"d.xhtml\\">page 4</a>
                  </li>
                </ol>
              </li>
            </ol>
          </nav>
        </body>
      </html>
      ",
        "_meta": {
          "language": "en",
          "title": "Nav",
        },
        "_properties": undefined,
        "file": "nav.xhtml",
        "mediaType": "application/xhtml+xml",
      }
    `);
  });
});
