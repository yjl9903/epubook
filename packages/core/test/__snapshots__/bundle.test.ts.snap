// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Bundle Epub > generate toc 1`] = `
PackageDocument {
  "SpecVersion": "3.0",
  "_identifier": "12345",
  "_items": [
    HTML {
      "_properties": undefined,
      "content": "1",
      "file": "page1.xhtml",
      "mediaType": "application/xhtml+xml",
    },
    HTML {
      "_properties": undefined,
      "content": "2",
      "file": "page2.xhtml",
      "mediaType": "application/xhtml+xml",
    },
    HTML {
      "_properties": undefined,
      "content": "3",
      "file": "page3.xhtml",
      "mediaType": "application/xhtml+xml",
    },
    HTML {
      "_properties": undefined,
      "content": "4",
      "file": "page4.xhtml",
      "mediaType": "application/xhtml+xml",
    },
    HTML {
      "_properties": undefined,
      "content": "5",
      "file": "page5.xhtml",
      "mediaType": "application/xhtml+xml",
    },
  ],
  "_metadata": {
    "contributor": [],
    "coverage": "",
    "creator": {
      "fileAs": "XLor",
      "name": "XLor",
      "uid": "creator",
    },
    "date": 2023-02-01T11:00:00.000Z,
    "description": "for test usage",
    "format": "",
    "language": "zh-CN",
    "lastModified": 2023-02-26T11:00:00.000Z,
    "publisher": "",
    "relation": "",
    "rights": "",
    "source": "imagine",
    "subject": "",
    "title": "Test Book",
    "type": "",
  },
  "_spine": [
    ManifestItemRef {
      "_idref": "page1",
      "optional": {},
    },
    ManifestItemRef {
      "_idref": "page2",
      "optional": {},
    },
    ManifestItemRef {
      "_idref": "page3",
      "optional": {},
    },
    ManifestItemRef {
      "_idref": "page4",
      "optional": {},
    },
    ManifestItemRef {
      "_idref": "page5",
      "optional": {},
    },
  ],
  "_toc": Toc {
    "_builder": XHTMLBuilder {
      "_body": [
        {
          "attrs": {
            "epub:type": "toc",
          },
          "children": [
            {
              "attrs": {},
              "children": [
                "Toc",
              ],
              "tag": "h2",
            },
            {
              "attrs": {},
              "children": [
                {
                  "attrs": {},
                  "children": [
                    {
                      "attrs": {
                        "href": "page1.xhtml",
                      },
                      "children": [
                        "1",
                      ],
                      "tag": "a",
                    },
                  ],
                  "tag": "li",
                },
                {
                  "attrs": {},
                  "children": [
                    {
                      "attrs": {
                        "href": "page2.xhtml",
                      },
                      "children": [
                        "2",
                      ],
                      "tag": "a",
                    },
                  ],
                  "tag": "li",
                },
                {
                  "attrs": {},
                  "children": [
                    {
                      "attrs": {},
                      "children": [
                        "Sub",
                      ],
                      "tag": "span",
                    },
                    {
                      "attrs": {},
                      "children": [
                        {
                          "attrs": {},
                          "children": [
                            {
                              "attrs": {
                                "href": "page3.xhtml",
                              },
                              "children": [
                                "3",
                              ],
                              "tag": "a",
                            },
                          ],
                          "tag": "li",
                        },
                        {
                          "attrs": {},
                          "children": [
                            {
                              "attrs": {
                                "href": "page4.xhtml",
                              },
                              "children": [
                                "4",
                              ],
                              "tag": "a",
                            },
                          ],
                          "tag": "li",
                        },
                      ],
                      "tag": "ol",
                    },
                  ],
                  "tag": "li",
                },
                {
                  "attrs": {},
                  "children": [
                    {
                      "attrs": {
                        "href": "page5.xhtml",
                      },
                      "children": [
                        "5",
                      ],
                      "tag": "a",
                    },
                  ],
                  "tag": "li",
                },
              ],
              "tag": "ol",
            },
          ],
          "tag": "nav",
        },
      ],
      "_filename": "nav.xhtml",
      "_head": [],
      "meta": {
        "language": "zh-CN",
        "title": "Toc",
      },
    },
    "_properties": "nav",
    "file": "nav.xhtml",
    "mediaType": "application/xhtml+xml",
  },
  "_uniqueIdentifier": "book-id",
  "file": "OEBPS/content.opf",
}
`;
